### PR TITLE
Fixes to get batch pipeline working and batch run results

### PR DESCRIPTION
**Pull request by LinfanS**
Hi all,

This pull request is for changes that were made to get the batch_pipeline working consistently. The batch pipeline was successfully run and I've included the log file, text file of bad GPT responses and a pg_dump file of the database state following the batch run should we find ourselves in a situation where we need to reset the database or it dies.

Summary of changes:

- Added `batch_pipeline.log` file that stores all the console outputs from the batch run. (bonus points for finding the times when it crashed and I was asleep).

- Added `court_transcript_batch_backup.sql` which is created from pg_dumps allowing us to restore the db state should we ever need to.

- Switched progress tracking from track to progress from rich.progress

- Added logging functionality for the batch pipeline and also in transform to log token cost and invalid GPT responses. I have removed all print statements as a result of this and instead facilitating both writing to file and console. I plan to later replicate something similar for the live_pipeline.

- Improved GPT prompting to handle cases with no claimant/defendant/lawyers/law firms and correctly form the participant dictionary with Nonetype values instead of just a Nonetype value.

- Relaxed timeout duration in `extract.py` as some pages take a bit longer for a response e.g. longer case files.

- Fixed an issue in `load.py` where replacing tags with synonyms would add trailing letters to tags such as `claimant` -> `claimantantantantantant`. @annac02 This might require you to update tests that use `replace_word_in_list`

- Further relaxed VARCHAR constraints for several tables in `schema.sql` and by extension `reset.sh`

- Added functionality to write invalid gpt responses to a text file called `invalid_gpt_responses.txt`, added additional check so that this will only happen for the batch_pipeline and not the live pipeline

- Updated GPT response validation functions in `transform.py` to add a function that checks the `first_side` and `second_side` dicts from the response are shaped correctly including nested elements and not just the outer data type. This is the most critical change that was required to get the batch pipeline to work.

- Fixed a subtle yet equally critical issue associated with serialisation of a python dict to a string and back between the Redis cache in `batch_pipeline.py`  as demonstrated below, have replaced it using literal_eval instead.
```python
import json
from ast import literal_eval


my_dict = {None: {None: None}}

print(json.loads(json.dumps(my_dict)))  # {'null': {'null': None}}

print(literal_eval(str(my_dict)))  # {None: {None: None}}

```

### CODE REVIEW

File: pipeline/batch_pipeline.py [hunk 1]
```diff
@@ -17,45 +18,70 @@
...
-                )
-
-        table_data = assemble_data(gpt_response)
-        conn = get_connection()
-        insert_to_database(conn, table_data)
-        print(f"Page {page_num} inserted to database, {count} records in total")
-
-    return "Data inserted to database"
+        message = "No new data to insert, exiting"
+        logger.info(message)
```
Comment (MilesDrabwell) [inline pipeline/batch_pipeline.py:43]: Can you not just put the string directly into logger.info?

File: pipeline/batch_pipeline.py [hunk 2]
```diff
@@ -17,45 +18,70 @@
...
+        return None
+
+    count = 0
```
Comment (MilesDrabwell) [inline pipeline/batch_pipeline.py:46]: Maybe a more specific variable name to show what this is counting?

File: pipeline/transform.py [hunk 3]
```diff
@@ -51,10 +53,31 @@ def get_summary(prompt: str, transcript: str) -> ChatCompletion:
...
+    usage = completion.usage
+    completion_tokens = getattr(usage, "completion_tokens", 0)
+    prompt_tokens = getattr(usage, "prompt_tokens", 0)
+    total_tokens = getattr(usage, "total_tokens", 0)
+    cost = [completion_tokens, prompt_tokens, total_tokens]
+    logger.info("GPT-4o-mini usage cost: %s", cost)
     return completion
 
 
+def is_valid_participant(data: dict) -> bool:
```
Comment (MilesDrabwell) [inline pipeline/transform.py:65]: Maybe the data parameter could have a better variable name?

**Review by MilesDrabwell — APPROVED** [review]
Overall great changes.

File: pipeline/transform.py [hunk 4]
```diff
@@ -51,10 +53,31 @@ def get_summary(prompt: str, transcript: str) -> ChatCompletion:
         ],
         temperature=0.1,  # low temperature to ensure the model doesn't diverge from the prompt
     )
-    print(completion.usage)
+    usage = completion.usage
+    completion_tokens = getattr(usage, "completion_tokens", 0)
+    prompt_tokens = getattr(usage, "prompt_tokens", 0)
+    total_tokens = getattr(usage, "total_tokens", 0)
+    cost = [completion_tokens, prompt_tokens, total_tokens]
+    logger.info("GPT-4o-mini usage cost: %s", cost)
```
Comment (annac02) [inline pipeline/transform.py:61]: nice - good thing to know and keep track of

File: pipeline/reset.sh
Comment (annac02) [inline pipeline/reset.sh:1]: ✅

File: pipeline/schema.sql [hunk 5]
```diff
@@ -57,13 +57,13 @@ CREATE TABLE lawyer (
...
-    court_case_id VARCHAR(50) UNIQUE NOT NULL,
+    court_case_id VARCHAR(250) UNIQUE NOT NULL,
     summary TEXT,
     verdict_id INT,
-    title VARCHAR(150),
+    title VARCHAR(512),
     court_date DATE,
-    case_number VARCHAR(50) UNIQUE NOT NULL,
-    case_url VARCHAR(200),
+    case_number VARCHAR(250),
```
Comment (annac02) [inline pipeline/schema.sql:65]: same as reset - why not 'not null'?

File: pipeline/schema.sql
Comment (annac02) [inline pipeline/schema.sql:1]: ✅

File: pipeline/invalid_gpt_responses.txt
Comment (annac02) [inline pipeline/invalid_gpt_responses.txt:1]: considerin it ran for a couple of days, 18 is not bad :) nice

File: pipeline/batch_pipeline.py [hunk 6]
```diff
@@ -17,45 +18,70 @@
...
+                        mapping={"case_details": str(response)},
+                    )
+                    gpt_response.append(response)
+                    message = (
+                        f"Cache miss for {court_case_citation}, "
+                        "GPT data fetched and stored in Redis."
+                    )
+                logger.info(message)
+
+            insert_to_database(get_connection(), assemble_data(gpt_response, True))
```
Comment (alabbahs) [inline pipeline/batch_pipeline.py:77]: Is this main function not insanely large, perhaps break it down into smaller functions that complete the different parts such as one part that processes the website pages, another that handles the cache and a final function that inserts it?

**Review by alabbahs — COMMENTED** [review]
Extremely hard work and diligence, shows care and attention to detail as well as the ability to polish the batch pipeline that was already written. Great work.

**Review by alabbahs — APPROVED** [review]

File: pipeline/batch_pipeline.py (hunk 1 above)
Comment (LinfanS) [reply pipeline/batch_pipeline.py:43]: Yes, that's completely true, fixed.

File: pipeline/batch_pipeline.py (hunk 2 above)
Comment (LinfanS) [reply pipeline/batch_pipeline.py:46]: Renamed to `cases_count`, thanks Miles

File: pipeline/transform.py (hunk 3 above)
Comment (LinfanS) [reply pipeline/transform.py:65]: Renamed to participant_dict. Thanks Miles.

File: pipeline/batch_pipeline.py [hunk 7]
```diff
@@ -17,45 +18,70 @@
...
+    file_handler = logging.FileHandler("batch_pipeline.log")
+    console_handler = logging.StreamHandler()
+    file_handler.setLevel(logging.INFO)
+    console_handler.setLevel(logging.INFO)
+    formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
+    file_handler.setFormatter(formatter)
+    console_handler.setFormatter(formatter)
+    logger.addHandler(file_handler)
+    logger.addHandler(console_handler)
+    return logger
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/batch_pipeline.py:34]: :)

File: pipeline/court_transcript_batch_backup.sql
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/court_transcript_batch_backup.sql:1]: Likely will back this up onto the S3 bucket as a safety measure, thanks for reminding.

File: pipeline/batch_pipeline.py [hunk 8]
```diff
@@ -1,9 +1,10 @@
 """Python script to run the batch pipeline locally to insert past 
 court cases to the database and cache GPT data in Redis"""
 
-import json
+from ast import literal_eval
+import logging
 import redis
-from rich.progress import track
+from rich.progress import Progress
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/batch_pipeline.py:7]: Progress generally more customisable and powerful than track at the expense of a little more configuring, still was not perfect in this case as probably needed to add another one for the inner loop to get times work but the percentage tracking was helpful.

File: pipeline/reset.sh [hunk 9]
```diff
@@ -62,13 +62,13 @@ CREATE TABLE lawyer (
...
-    court_case_id VARCHAR(50) UNIQUE NOT NULL,
+    court_case_id VARCHAR(250) UNIQUE NOT NULL,
     summary TEXT,
     verdict_id INT,
-    title VARCHAR(200),
+    title VARCHAR(512),
     court_date DATE,
-    case_number VARCHAR(50) UNIQUE NOT NULL,
-    case_url VARCHAR(250),
+    case_number VARCHAR(250),
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/reset.sh:70]: If I recall there was a conflict with multiple cases that couldn't find a case number and hence a duplicate value.

File: pipeline/transform.py [hunk 10]
```diff
@@ -51,10 +53,31 @@ def get_summary(prompt: str, transcript: str) -> ChatCompletion:
...
+            return False
+        if not isinstance(value, dict):
+            return False
+        for sub_key, sub_value in value.items():
+            if not isinstance(sub_key, (str, type(None))):
+                return False
+            if not isinstance(sub_value, (str, type(None))):
+                return False
+
+    return True
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/transform.py:78]: Thanks

File: pipeline/schema.sql (hunk 5 above)
Comment (LinfanS) [reply pipeline/schema.sql:65]: As I believe at one point, there may have been null values for case_number (if GPT couldn't find one) for the odd case or two that was causing the batch_pipeline to crash.

File: pipeline/prompts.py [hunk 11]
```diff
@@ -14,16 +14,16 @@
     I only want an unnamed python dictionary that gives the key information of the transcript as such:
 
     {
-    "case_number": "EA-2022-000822-AS",
+    "case_number": "EA-2022-000822-AS", might also be called claim number in the text
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/prompts.py:17]: Thanks Hamoodi

File: pipeline/transform.py [hunk 12]
```diff
@@ -51,10 +53,31 @@ def get_summary(prompt: str, transcript: str) -> ChatCompletion:
...
+    cost = [completion_tokens, prompt_tokens, total_tokens]
+    logger.info("GPT-4o-mini usage cost: %s", cost)
     return completion
 
 
+def is_valid_participant(data: dict) -> bool:
+    """Check if the participant dict is valid"""
+    for key, value in data.items():
+        if not isinstance(key, (str, type(None))):
+            return False
```
[the remark this answers is not in the document]
Comment (LinfanS) [reply pipeline/transform.py:69]: I'll look into this, thanks Hamoodi.

File: pipeline/batch_pipeline.py (hunk 6 above)
Comment (LinfanS) [reply pipeline/batch_pipeline.py:77]: Yes, this is a good point, the main point for keeping it together at the moment is my understanding is that you have to keep passing the logger around as a function argument in the subsequent functions - an issue I encountered when doing this for the live_pipeline, definitely something we should double back around too if we have time.

**Pull request merged**